File: src/main/java/kr/co/lunatalk/domain/productlike/repository/ProductLikeRepositoryImpl.java
package kr.co.lunatalk.domain.productlike.repository;

import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Repository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static kr.co.lunatalk.domain.productlike.domain.QProductLike.productLike;

@Repository
@RequiredArgsConstructor
public class ProductLikeRepositoryImpl implements ProductLikeRepositoryCustom {

	private final JPAQueryFactory queryFactory;

	@Override
	public Long countByProductId(Long productId) {
		return queryFactory
			.select(productLike.count())
			.from(productLike)
			.where(productIdEq(productId))
			.fetchOne();
	}

	@Override
	public Map<Long, Long> countByProductIds(List<Long> productIds) {
		if (productIds.isEmpty()) {
			return new HashMap<>();
		}

		List<CountResult> results = queryFactory
			.select(productLike.product.id, productLike.count())
			.from(productLike)
			.where(productLike.product.id.in(productIds))
			.groupBy(productLike.product.id)
			.fetch()
			.stream()
			.map(tuple -> new CountResult(
				tuple.get(productLike.product.id),
				tuple.get(productLike.count())
			))
			.toList();

		Map<Long, Long> countMap = new HashMap<>();
		for (Long productId : productIds) {
			countMap.put(productId, 0L);
		}
		for (CountResult result : results) {
			countMap.put(result.productId, result.count);
		}
		return countMap;
	}

	@Override
	public Map<Long, Boolean> existsByMemberIdAndProductIds(Long memberId, List<Long> productIds) {
		List<Long> likedProductIds = queryFactory
			.select(productLike.product.id)
			.from(productLike)
			.where(
				productLike.member.id.eq(memberId),
				productLike.product.id.in(productIds)
			)
			.fetch();

		Map<Long, Boolean> resultMap = new HashMap<>();
		for (Long productId : productIds) {
			resultMap.put(productId, likedProductIds.contains(productId));
		}
		return resultMap;
	}

	@Override
	public Page<Long> findLikedProductIdsByMemberId(Long memberId, Pageable pageable) {
		List<Long> likedProductIds = queryFactory
			.select(productLike.product.id)
			.from(productLike)
			.where(productLike.member.id.eq(memberId))
			.orderBy(productLike.createdAt.desc(), productLike.id.desc())
			.offset(pageable.getOffset())
			.limit(pageable.getPageSize())
			.fetch();

		Long total = Optional.ofNullable(
			queryFactory
				.select(productLike.count())
				.from(productLike)
				.where(productLike.member.id.eq(memberId))
				.fetchOne()
		).orElse(0L);

		return new PageImpl<>(likedProductIds, pageable, total);
	}

	private BooleanExpression productIdEq(Long productId) {
		return productId != null ? productLike.product.id.eq(productId) : null;
	}

	private record CountResult(Long productId, Long count) {
	}
}

File: lunatalk-core/src/main/kotlin/kr/co/lunatalk/domain/productlike/repository/ProductLikeRepositoryImpl.kt
package kr.co.lunatalk.domain.productlike.repository

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import kr.co.lunatalk.domain.productlike.domain.QProductLike.productLike
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class ProductLikeRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : ProductLikeRepositoryCustom {

    override fun countByProductId(productId: Long): Long? {
        return queryFactory
            .select(productLike.count())
            .from(productLike)
            .where(productIdEq(productId))
            .fetchOne()
    }

    override fun countByProductIds(productIds: List<Long>): Map<Long, Long> {
        if (productIds.isEmpty()) {
            return HashMap()
        }

        val results = queryFactory
            .select(productLike.product.id, productLike.count())
            .from(productLike)
            .where(productLike.product.id.`in`(productIds))
            .groupBy(productLike.product.id)
            .fetch()
            .map { tuple ->
                CountResult(
                    tuple.get(productLike.product.id)!!,
                    tuple.get(productLike.count())!!
                )
            }

        val countMap = HashMap<Long, Long>()
        for (productId in productIds) {
            countMap[productId] = 0L
        }
        for (result in results) {
            countMap[result.productId] = result.count
        }
        return countMap
    }

    override fun existsByMemberIdAndProductIds(memberId: Long, productIds: List<Long>): Map<Long, Boolean> {
        val likedProductIds = queryFactory
            .select(productLike.product.id)
            .from(productLike)
            .where(
                productLike.member.id.eq(memberId),
                productLike.product.id.`in`(productIds)
            )
            .fetch()

        val resultMap = HashMap<Long, Boolean>()
        for (productId in productIds) {
            resultMap[productId] = likedProductIds.contains(productId)
        }
        return resultMap
    }

    override fun findLikedProductIdsByMemberId(memberId: Long, pageable: Pageable): Page<Long> {
        val likedProductIds = queryFactory
            .select(productLike.product.id)
            .from(productLike)
            .where(productLike.member.id.eq(memberId))
            .orderBy(productLike.createdAt.desc(), productLike.id.desc())
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .fetch()

        val total = Optional.ofNullable(
            queryFactory
                .select(productLike.count())
                .from(productLike)
                .where(productLike.member.id.eq(memberId))
                .fetchOne()
        ).orElse(0L)

        return PageImpl(likedProductIds, pageable, total)
    }

    private fun productIdEq(productId: Long?): BooleanExpression? {
        return productId?.let { productLike.product.id.eq(it) }
    }

    private data class CountResult(val productId: Long, val count: Long)
}
